File: src/main/java/semicontinuity/idea/code/analyzer/graph/viewModel/vanilla/VFactory.java
package semicontinuity.idea.code.analyzer.graph.viewModel.vanilla;

import java.util.List;

import semicontinuity.idea.code.analyzer.graph.viewModel.Factory;

public class VFactory<PAYLOAD> implements Factory<
        PAYLOAD,
        VComponent,
        VIndependentComponents,
        VNode<PAYLOAD>,
        VSplit<PAYLOAD>,
        VLayer
        > {

    @Override
    public VIndependentComponents newIndependentComponents(List<? extends VComponent> components) {
        return new VIndependentComponents(components);
    }

    @Override
    public VNode<PAYLOAD> newNode(PAYLOAD payload) {
        return new VNode<>(payload);
    }

    @Override
    public VSplit<PAYLOAD> newSplit(List<VNode<PAYLOAD>> items, VComponent subLayer) {
        return new VSplit<>(items, subLayer);
    }

    @Override
    public VLayer newLayer(VComponent directDeps, VComponent sharedDeps) {
        return new VLayer(directDeps, sharedDeps);
    }
}

File: src/main/java/semicontinuity/idea/code/analyzer/graph/viewModel/vanilla/VFactory.kt
package semicontinuity.idea.code.analyzer.graph.viewModel.vanilla

import semicontinuity.idea.code.analyzer.graph.viewModel.Factory

class VFactory<PAYLOAD> :
    Factory<PAYLOAD, VComponent?, VIndependentComponents?, VMember<PAYLOAD>?, VSplit<PAYLOAD>?, VLayer?, VSplit0> {

    override fun newIndependentComponents(components: List<VComponent?>): VIndependentComponents {
        return VIndependentComponents(components)
    }

    override fun newVertex(payload: PAYLOAD): VMember<PAYLOAD> {
        return VMember(payload)
    }

    override fun newSplit(left: VComponent?, right: VComponent?): VSplit0 {
        return VSplit0(left, right)
    }

    override fun newSplit(items: List<VMember<PAYLOAD>?>, subLayer: VComponent?): VSplit<PAYLOAD> {
        return VSplit(items, subLayer)
    }

    override fun newLayer(directDeps: VComponent?, sharedDeps: VComponent?): VLayer {
        return VLayer(directDeps, sharedDeps)
    }
}
